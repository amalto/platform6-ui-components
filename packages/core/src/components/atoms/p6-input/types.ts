export enum P6InputType {
  email = 0,
  number = 1,
  password = 2,
  search = 3,
  tel = 4,
  text = 5,
  url = 6,
}
