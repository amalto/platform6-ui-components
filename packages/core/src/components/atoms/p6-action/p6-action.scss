$button-color: $text-strong !default;
$button-background-color: $scheme-main !default;
$button-family: false !default;
$button-border-width: $control-border-width !default;
$action-padding-vertical: calc(0.1em - #{$button-border-width}) !default;
$action-padding-horizontal: 0.2em !default;
$button-hover-color: $link-hover !default;
$button-focus-color: $link-focus !default;
$button-active-color: $link-active !default;
$button-disabled-background-color: $scheme-main !default;
$button-disabled-opacity: 0.5 !default;

@import 'src/styles/atoms/control';

:host {
  display: inline-flex;
  align-items: center;
  justify-content: center;

  button {
    @include control;
    @include unselectable;

    display: inline-flex;
    background-color: $button-background-color;
    border-width: $button-border-width;
    color: $button-color;
    cursor: pointer;
    font-family: $body-family;

    @if $button-family {
      font-family: $button-family;
    }

    justify-content: center;
    padding: $action-padding-vertical $action-padding-horizontal;
    text-align: center;
    white-space: nowrap;
    width: auto;
    height: auto;

    ::slotted(.icon:first-child:last-child),
    ::slotted(p6-icon:first-child:last-child) {
      $margin: calc(#{-1 / 2 * $action-padding-horizontal} - #{$button-border-width});

      margin-left: $margin;
      margin-right: $margin;
    }

    ::slotted(p6-icon:active),
    ::slotted(.icon:active) {
      background-attachment: grey;
      outline-color: transparent;
    }

    // States
    &:hover {
      color: $button-hover-color;
    }
    &:focus {
      color: $button-focus-color;
    }
    &:active {
      color: $button-active-color;
    }

    // Colors
    @each $name, $pair in $colors {
      $color: nth($pair, 1);
      $color-invert: nth($pair, 2);

      &.is-#{$name} {
        background-color: transparent;
        border-color: transparent; // $color;
        color: $color;

        &:hover,
        &:focus {
          background-color: transparent;
          border-color: transparent;
          color: bulmaDarken($color, 2.5%);
        }

        &:active {
          background-color: transparent;
          border-color: transparent;
          color: bulmaDarken($color, 5%);
        }

        &[disabled],
        fieldset[disabled] :host(.is-#{$name}) {
          background-color: transparent;
          border-color: transparent;
          box-shadow: none;
          color: $color;
        }

        &.is-loading {
          &::after {
            border-color: transparent transparent $color $color !important;
          }

          &:hover::after,
          &:focus::after {
            border-color: transparent transparent bulmaDarken($color, 2.5%) bulmaDarken($color, 2.5%) !important;
          }
        }
      }
    }

    // Size
    &.is-small {
      border-radius: $radius-small;
      font-size: $size-small;
    }
    &.is-normal {
      font-size: $size-normal;
    }

    &.is-loading {
      color: transparent !important;
      pointer-events: none;
      &::after {
        @include loader;
        @include center(1em);

        position: absolute !important;
      }
    }

    // Modifiers
    &[disabled],
    fieldset[disabled] :host {
      background-color: $button-disabled-background-color;
      border-color: transparent;
      opacity: $button-disabled-opacity;
    }
  }
}
