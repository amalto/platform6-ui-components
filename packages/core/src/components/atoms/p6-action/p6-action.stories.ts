import { Components } from '../../../components';
import { ComponentProps, getElement, makeModeStory, makeSizeStory, makeStory, Props } from '../../../shared/storybook';
import { Mode, Size } from '../../../shared/types';

const component = 'p6-action';

export default {
  title: 'Atoms/Action',
  component,
};

const componentProps: ComponentProps = ['mode', 'waiting', 'disabled', 'size'];

const getStoryField = (props?: Props<Components.P6Action>): HTMLElement => {
  const size = props?.size !== undefined ? props.size : Size.normal;
  return getElement(
    component,
    getElement('p6-icon', [], {
      size,
      name: 'home',
    }),
    props,
  );
};

export const Default = makeStory<{
  size: Size;
  mode: Mode;
  disabled: boolean;
  waiting: boolean;
}>({
  componentProps,
  args: {
    size: Size.normal,
    mode: Mode.default,
    disabled: false,
    waiting: false,
  },
  builder: args => getStoryField(args),
});

export const Modes = makeModeStory({
  componentProps,
  builder: ({ value }) =>
    getStoryField({
      mode: value,
    }),
});

export const Disabled = makeStory<{ disabled: boolean }>({
  componentProps,
  args: { disabled: true },
  builder: (props): HTMLElement => getStoryField(props),
});

export const Waiting = makeStory<{ waiting: boolean }>({
  componentProps,
  args: { waiting: true },
  builder: (props): HTMLElement => getStoryField(props),
});

export const Sizes = makeSizeStory({
  componentProps,
  builder: ({ value }) =>
    getStoryField({
      size: value,
    }),
});
