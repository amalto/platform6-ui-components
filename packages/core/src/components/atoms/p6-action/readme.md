# p6-action

<!-- Auto Generated Below -->

## Properties

| Property   | Attribute  | Description                                          | Type                                                                                       | Default        |
| ---------- | ---------- | ---------------------------------------------------- | ------------------------------------------------------------------------------------------ | -------------- |
| `disabled` | `disabled` | If `true`, the user cannot interact with the Action. | `boolean`                                                                                  | `false`        |
| `mode`     | `mode`     | set the mode of the action                           | `Mode.danger \| Mode.default \| Mode.info \| Mode.primary \| Mode.success \| Mode.warning` | `Mode.default` |
| `size`     | `size`     | set the size of the action                           | `Size.normal \| Size.small`                                                                | `Size.normal`  |
| `waiting`  | `waiting`  | If set, shows a waiting/busy indicator               | `boolean`                                                                                  | `false`        |

## Dependencies

### Used by

- [p6-grid-header-menu](../../organisms/p6-grid/components/p6-grid-header-menu)
- [p6-grid-row-actions](../../templates/p6-grid-row-actions)
- [p6-panel](../../molecules/p6-panel)
- [p6-service-panel](../../templates/p6-service-panel)
- [p6-translation](../../molecules/p6-translation)
- [p6-translation-line](../../molecules/p6-translation/components/p6-translation-line)

### Graph

```mermaid
graph TD;
  p6-grid-header-menu --> p6-action
  p6-grid-row-actions --> p6-action
  p6-panel --> p6-action
  p6-service-panel --> p6-action
  p6-translation --> p6-action
  p6-translation-line --> p6-action
  style p6-action fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
