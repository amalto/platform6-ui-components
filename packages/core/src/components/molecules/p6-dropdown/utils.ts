export function addDropdownClass(child: Element): Element {
  if (child.nodeName === 'HR') {
    child.classList.add('dropdown-divider');
    return child;
  }

  child.classList.add('dropdown-item');
  return child;
}
