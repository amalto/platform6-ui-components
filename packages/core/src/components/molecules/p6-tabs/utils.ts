export function getTabId(id: string): string {
  return `${id}-tab`;
}

export function isTabValid(tab: Element): tab is HTMLElement {
  return 'title' in tab && 'id' in tab;
}
