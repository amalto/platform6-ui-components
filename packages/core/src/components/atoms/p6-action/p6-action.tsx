import { Component, ComponentInterface, Element, h, Host, JSX, Prop } from '@stencil/core';
import { Mode, Size } from '../../../shared/types';
import { getModeClass, getSizeClass } from '../../../utils/classes';

@Component({
  tag: 'p6-action',
  styleUrl: 'p6-action.scss',
  shadow: true,
})
export class P6Action implements ComponentInterface {
  @Element() host!: HTMLP6ActionElement;

  /**
   * set the mode of the action
   */
  @Prop() mode: Mode = Mode.default;

  /**
   * If set, shows a waiting/busy indicator
   */
  @Prop() waiting = false;

  /**
   * If `true`, the user cannot interact with the Action.
   */
  @Prop() disabled = false;

  /**
   * set the size of the action
   */
  @Prop() size: Size = Size.normal;

  render(): JSX.Element {
    const classes = {
      'is-loading': !!this.waiting,
      ...getModeClass(this.mode),
      ...getSizeClass(this.size),
    };

    return (
      <Host aria-disabled={this.disabled ? 'true' : null}>
        <button type="button" disabled={this.disabled} class={classes}>
          <slot />
        </button>
      </Host>
    );
  }
}
