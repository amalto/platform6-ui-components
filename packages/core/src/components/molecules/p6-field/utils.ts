import { isP6Control, isP6NativeControl, P6Control, P6NativeControl } from '../../../shared/form/control';
import { isInDefaultSlot } from '../../../utils/component';

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export function isControl(elmt: unknown): elmt is P6Control<unknown> | P6NativeControl {
  return isP6Control(elmt) || isP6NativeControl(elmt);
}

export function getControl(children: HTMLCollection): P6Control<unknown> | P6NativeControl | undefined {
  return Array.from(children).filter(isInDefaultSlot).find(isControl) as P6Control<unknown> | P6NativeControl | undefined;
}
