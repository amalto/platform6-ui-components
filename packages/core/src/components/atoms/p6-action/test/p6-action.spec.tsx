import { newSpecPage } from '@stencil/core/testing';
import { enumArrayEntryToArray } from '../../../../shared/test/utils';
import { Mode, modes, Size, sizes } from '../../../../shared/types';
import { P6Action } from '../p6-action';

describe('p6-action', () => {
  describe('mode', () => {
    it.each(enumArrayEntryToArray(modes))('has the class "is-%s" when mode is "%s"', async (_, mode) => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action mode=${mode}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('does not add class when mode is default', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action mode=${Mode.default}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('does not add class when mode is undefined', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action mode=${undefined}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('waiting', () => {
    it('has a "is-loading" class when waiting is true', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action waiting="true"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "is-loading" class when waiting is false', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action waiting="false"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('disabled', () => {
    it('has a "disabled" attribute on the button when disabled is true', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action disabled="true"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "disabled" attribute on the button when disabled is false', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action disabled="false"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('size', () => {
    it.each(enumArrayEntryToArray(sizes))('has the class "is-%s" when size is "%s"', async (_, size) => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action size=${size}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('does not add class when size is normal', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action size=${Size.normal}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
});
