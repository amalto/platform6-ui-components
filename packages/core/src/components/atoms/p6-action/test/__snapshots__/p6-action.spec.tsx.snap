// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-action disabled has a "disabled" attribute on the button when disabled is true 1`] = `
<p6-action aria-disabled="true" disabled="true">
  <mock:shadow-root>
    <button disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action disabled has no "disabled" attribute on the button when disabled is false 1`] = `
<p6-action disabled="false">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode does not add class when mode is default 1`] = `
<p6-action mode="2">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode does not add class when mode is undefined 1`] = `
<p6-action mode="undefined">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-danger" when mode is "0" 1`] = `
<p6-action mode="0">
  <mock:shadow-root>
    <button class="is-danger" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-default" when mode is "2" 1`] = `
<p6-action mode="2">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-info" when mode is "3" 1`] = `
<p6-action mode="3">
  <mock:shadow-root>
    <button class="is-info" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-primary" when mode is "5" 1`] = `
<p6-action mode="5">
  <mock:shadow-root>
    <button class="is-primary" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-success" when mode is "4" 1`] = `
<p6-action mode="4">
  <mock:shadow-root>
    <button class="is-success" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-warning" when mode is "1" 1`] = `
<p6-action mode="1">
  <mock:shadow-root>
    <button class="is-warning" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size does not add class when size is normal 1`] = `
<p6-action size="1">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size has the class "is-normal" when size is "1" 1`] = `
<p6-action size="1">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size has the class "is-small" when size is "0" 1`] = `
<p6-action size="0">
  <mock:shadow-root>
    <button class="is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action waiting has a "is-loading" class when waiting is true 1`] = `
<p6-action waiting="true">
  <mock:shadow-root>
    <button class="is-loading" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action waiting has no "is-loading" class when waiting is false 1`] = `
<p6-action waiting="false">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;
